share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml

```diff
@@ -0,0 +1,97 @@
+spack:
+  view: false
+  packages:
+    all:
+      require: target=ppc64le
+      variants: ~cuda~rocm
+
+  specs:
+    # Horovod
+    # Still depends on py-tensorboard-data-server even when not using the tensorflow framework.
+    #- py-horovod
+
+    # Hugging Face
+    - py-transformers
+
+    # JAX
+    - py-jax
+    - py-jaxlib
+
+    # Keras
+    # Fails to install bazel and py-tensorboard-data-server
+    #- py-keras
+    - py-keras-applications
+    - py-keras-preprocessing
+    - py-keras2onnx
+
+    # MXNet
+    - mxnet
+
+    # PyTorch
+    - py-botorch
+    - py-efficientnet-pytorch
+    - py-gpytorch
+    - py-kornia
+    - py-lightning
+    - py-pytorch-gradual-warmup-lr
+    - py-pytorch-lightning
+    - py-segmentation-models-pytorch
+    - py-timm
+    - py-torch
+    - py-torch-cluster
+    - py-torch-geometric
+    - py-torch-nvidia-apex
+    - py-torch-scatter
+    - py-torch-sparse
+    - py-torch-spline-conv
+    - py-torchaudio
+    - py-torchdata
+    - py-torchfile
+    - py-torchgeo
+    - py-torchmetrics
+    - py-torchtext
+    - py-torchvision
+    - py-vector-quantize-pytorch
+
+    # scikit-learn
+    - py-scikit-learn
+    - py-scikit-learn-extra
+
+    # TensorBoard
+      # py-tensorboard-data-server fails in CI when compiling `ring v0.16.20` as part
+      # of the rust-cargo dependency builds.
+      #- py-tensorboard
+      #- py-tensorboard-data-server
+      #- py-tensorboard-plugin-wit
+      #- py-tensorboardx
+
+    # TensorFlow
+    # See py-tensorboard-data-server failures
+    #- py-tensorflow
+    #- py-tensorflow-datasets
+    #- py-tensorflow-estimator
+    #- py-tensorflow-hub
+    #- py-tensorflow-metadata
+    #- py-tensorflow-probability
+
+    # XGBoost
+    - py-xgboost
+    # - r-xgboost
+    - xgboost
+
+  mirrors:
+    mirror: s3://spack-binaries/develop/ml-linux-ppc64le-cpu
+
+  ci:
+    pipeline-gen:
+    - build-job:
+        image:
+          name: ecpe4s/ubuntu20.04-runner-ppc64le:2023-01-01
+          entrypoint: ['']
+    - build-job:
+        image:
+          name: ecpe4s/rhel7-runner-ppc64le:2023-01-01
+          entrypoint: ['']
```

Comment:
Did you mean to have some submapping here, rather than just `build-job`?  If this was working, I'm not sure how.  Given you were trying to have two `spack.yaml` for the four pipelines, you may have been able to do something like this instead:

Example:

```yaml
...
 - match_behavior: first
    submapping:
    - match:
      - os=ubuntu20.04
        build-job:
          image:
            name: ecpe4s/ubuntu20.04-runner-ppc64le:2023-01-01
            entrypoint: ['']
```

But that's irrelevant if you change to one stack per pipeline.  If you wait until #39939 is merged, then I think you could stick with only two `spack.yaml` for your four pipelines, and the above example would be how you map build jobs to docker image.

cc @kwryankrattiger

Reply:
Not sure what the issue is here.  I want each stack to build on both ubuntu and rhel8.  In yaml, the `pipeline-gen` key takes a list, so I just put two elements in that list.  You can see from the CI build that this properly broke out the pipeline into separate jobs, each utilizing the appropriate container.

I am planning to wait for #39939 to be merged to avoid breaking this into multiple files only to be able to deduplicate after that PR goes through.

Reply:
It looks to me like it didn't break them out properly.  This [job](https://gitlab.spack.io/spack/spack/-/jobs/8227536) was supposed to be a rhel8, but ran on the ubuntu container.

Reply:
Interesting indeed.  The container does seem to indicate ubuntu for the rhel job.  But then the `uname -a` indicates a rhel OS.  Presumably that's the host, so fair enough.  The `spack arch` output seems to corroborate the ubuntu OS.  But then the `gnuconfig` dependency installation along with the libiconv installation directories indicate a rhel OS.  I guess that's due to the generate builds occurring with the correct container.  I just pushed a commit that tries to be more explicit with which container image should be used for the CI jobs.

Reply:
For reasons I cannot fathom, a CI jobs ending in `-build` will not accept an `image` key in the `.gitlab-ci.yaml` file.  Likewise, it can "extend" another job that does take an `image` key, but does not honor that same `image` when building.  I can try modifying the `spack.yaml` files according to your above example, but without proper documentation, I'll be doing a good bit of guess and check with CI parsing to try and get it right.

Reply:
> For reasons I cannot fathom, a CI jobs ending in -build will not accept an image key in the .gitlab-ci.yaml file.

Jobs ending in `-build`?  I guess the trigger jobs do seem to all end in `-build`, here's a [link](https://docs.gitlab.com/ee/ci/yaml/#trigger) to the gitlab docs describing what keys are valid in trigger jobs.

> but without proper documentation, I'll be doing a good bit of guess and check with CI parsing to try and get it right

This [bit](https://spack.readthedocs.io/en/latest/pipelines.html#ci-yaml) of the pipeline docs seems to illustrate an example of what you're trying to do.

Reply:
Forgot about this comment, but want to resurface this:

> I think testing both ubuntu and rhel is overkill right now. We don't even do that for x86_64. This is something worth adding someday but idk if we have the CI bandwidth to do both yet.

Reply:
I stand by my response to that comment: https://github.com/spack/spack/pull/39174#discussion_r1309191617

Reply:
The latest pipeline run https://gitlab.spack.io/spack/spack/-/pipelines/500642 is an example of exactly this happening.

Reply:
FYI, Spack is slowly moving in the direction of modeling glibc such that packages will no longer require _any_ system dependencies. Once this is done, Ubuntu and RHEL will be the same system, and there should be no real reason to test both.

Reply:
While I appreciate your optimism in this respect, the demonstrable differences between the two OSes means I'm going to have to see the two pipelines succeed side-by-side with a significant portion of the ML packages we're interested in before I'm willing to down sample to a single OS.